Fix issue 482

hra687261 opened the pull request:
Purification of functional array expressions, fixes issue #482.

hra687261 committed:
Extended purify_form (access to ('a, bool) farray)
Added the purification of the operands when the operator is an access to an functional array storing booleans

Gbury commented on src/lib/structures/expr.ml:
```diff
@@ -2433,6 +2433,30 @@ module Purification = struct
+         in which the stored values are booleans *)
+      | Sy.Op Get ->
+        begin match e.xs with
+          | [x; y] ->
```
Please give more informative names to these variables. For instance, to better distinguish which is the expression for the array, and which is the expression for the looked-up index.

hra687261 replied:
Ok, I replaced ```x``` with ```fa``` for functional array, and ```y``` with ```i``` for index.

Gbury commented on src/lib/structures/expr.ml:
```diff
@@ -2433,6 +2433,30 @@ module Purification = struct
+              if is_pure x then x
+              else
+                let x', lets = purify_term x SMap.empty in
+                mk_lifted x' lets
```
wouldn't it be better to lift the required let-bindings at the top ? i.e. something like: `let x', x'_lets = .... in let y', y'_lets = ... in mk_lifted ...` ?

hra687261 replied:
Yes, you're right. It's done.

Gbury reviewed (changes requested):
I tihk the fix is good, but I left a few comments. Also, while we're at it, could you change the `assert false` that are present in the purification functions into e.g. `failwith "unexpected expression in purification function"` so that the error message is less cryptic when such bugs occur again ?

hra687261 committed:
replaced evey "assert false" with a "failwith"
and made the lifting of let bindings be done at the top in farray access in purify_form

hra687261 committed:
Merge branch 'OCamlPro:next' into next-fix-482

hra687261 committed:
Merge branch 'OCamlPro:next' into next-fix-482

hra687261 committed:
Merge branch 'OCamlPro:next' into next-fix-482
